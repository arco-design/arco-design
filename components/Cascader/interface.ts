import { ReactNode, CSSProperties } from 'react';
import Store from './base/store';
import { TriggerProps } from '../Trigger';
import { SelectViewCommonProps } from '../_class/select-view';
import { NodeProps } from './base/node';
import { VirtualListProps } from '../_class/VirtualList';

/**
 * @title Cascader
 */
export interface CascaderProps<T = any>
  extends Omit<SelectViewCommonProps, 'allowCreate' | 'dragToSort'> {
  /**
   * @zh 选择框的默认值
   * @en Initial value
   */
  defaultValue?: (string | string[])[];

  /**
   * @zh 输入框的值
   * @en Input Value
   * @version 2.34.0
   */
  inputValue?: string;

  /**
   * @zh 选中值
   * @en To set value
   */
  value?: (string | string[])[];
  /**
   * @zh 级联数据
   * @en The data of options
   * @defaultValue []
   */
  options?: T[];
  trigger?: TriggerProps['trigger'];
  /**
   * @zh 展开下一级方式
   * @en Set the way to display the next level menu. One of hover and click
   * @defaultValue click
   */
  expandTrigger?: 'click' | 'hover';
  /**
   * @zh 每当选择的时候，值都会发生改变。多选时如果设置为`true`，会取消父子关系的关联。(默认只有在选择完成的时候，值才会更新)
   * @en Each selection will change value  if set to true.  when `mode=multiple`, child node and parent node will not affect each other.
   */
  changeOnSelect?: boolean;
  /**
   * @zh 是否在非动态加载时，选中项children为[]的时候渲染下一级节点。
   * @en Whether to render the next level node when the children of the selected option is an empty array
   */
  showEmptyChildren?: boolean;
  /**
   * @zh 是否在隐藏之后销毁DOM结构，默认为 `true`。如果是动态加载时，默认为`false`。
   * @en Whether destroy popup when hidden.
   */
  unmountOnExit?: boolean;
  /**
   * @zh 是否开启多选
   * @en Set mode
   */
  mode?: 'multiple';
  triggerProps?: Partial<TriggerProps>;
  /**
   * @zh 没有数据时显示的内容
   * @en The content to show when no result matches
   */
  notFoundContent?: ReactNode;
  /**
   * @zh 指定label，value，isLeaf，disabled，children 对应的字段
   * @en Custom field name for label, value, isLeaf, disabled and children
   * @defaultValue DefaultFieldNames
   */
  fieldNames?: FieldNamesType;
  /**
   * @zh 控制下拉框的展开收起
   * @en Whether the popup is visible
   */
  popupVisible?: boolean;
  /**
   * @zh 默认下拉框的展开收起状态
   * @en Whether the popup is visible by default
   */
  defaultPopupVisible?: boolean;
  /**
   * @zh
   * 定制回填方式 <br/> parent: 子节点都被选中时候返回父节点 <br/> child: 返回子节点
   * @en
   * Customize the return value  <br/> parent:
   * Only return the parent node when all child nodes are selected <br/> child: Return child nodes
   * @defaultValue child
   * @version 2.31.0
   */
  checkedStrategy?: 'parent' | 'child';
  /**
   * @zh 自定义下拉列表类名
   * @en Custom dropdown list classname
   * @version 2.35.0
   */
  dropdownMenuClassName?: string | string[];
  /**
   * @zh 菜单列样式
   * @en dropdown menu column style
   * @version 2.35.0
   */
  dropdownMenuColumnStyle?: CSSProperties;
  /**
   * @zh 传递虚拟滚动属性。开启虚拟滚动后，每列级联菜单的会存在默认宽度，可通过 `dropdownMenuColumnStyle` 进行样式调整
   * @en virtual list props. After virtual scrolling is enabled, there will be a default width for each column of cascading menus, which can be adjusted by `dropdownMenuColumnStyle`
   * @version 2.35.0
   */
  virtualListProps?: Pick<VirtualListProps<any>, 'threshold' | 'isStaticItemHeight'>;
  /**
   * @zh 自定义下拉菜单的展示。
   * @en Customize the popup menu.
   * @version 2.15.0
   */
  dropdownRender?: (menu: ReactNode) => ReactNode;
  /**
   * @zh 自定义下拉菜单每一列的展示。
   * @en Customize columns of the menu.
   * @version 2.15.0, `level` in 2.17.0
   */
  dropdownColumnRender?: (menu: ReactNode, level: number) => ReactNode;
  /**
   * @zh 默认搜索从 `label` 属性中进行关键字搜索。通过该方法可以自定义搜索逻辑
   * @en Customize the search logic.
   */
  filterOption?: (inputValue: string, option: NodeProps<T>) => boolean;
  /**
   * @zh 自定义展示 `option`
   * @en Custom rendering `option`
   */
  renderOption?: (option: NodeProps<T>, level: number) => ReactNode;
  /**
   * @zh 定义每一层级的 `footer`。参数：level: 当前层级, activeOption: 当前点击的节点。返回 `null` 不展示
   * @en Custom rendering the `footer` of each level menu.
   */
  renderFooter?: (level: number, activeOption: NodeProps<T> | null) => ReactNode;
  /**
   * @zh 格式化显示内容。
   * @en The return value will be displayed in the input box.
   */
  renderFormat?: (valueShow: any[]) => ReactNode;
  /**
   * @zh 搜索时的回调。(reason in `2.34.0`)
   * @en Callback when input changed.(reason in `2.34.0`)
   * @version 2.20.0
   */
  onSearch?: (inputValue: string, reason: InputValueChangeReason) => void;
  /**
   * @zh 点击选择框的回调。
   * @en Callback when finishing select.
   */
  onChange?: (
    value: (string | string[])[],
    selectedOptions,
    extra: { dropdownVisible?: boolean }
  ) => void;
  /**
   * @zh inputValue改变时的回调
   * @en Callback when inputValue change.
   * @version 2.34.0
   */
  onInputValueChange?: (inputValue: string, reason: InputValueChangeReason) => void;
  /**
   * @zh 弹出框挂在的父节点
   * @en ParentNode which the selector should be rendered to.
   */
  getPopupContainer?: (node: HTMLElement) => Element;
  /**
   * @zh 动态加载数据。pathValue: 当前选中项的路径 value； level: 选中项层级。
   * @en To load option lazily
   */
  loadMore?: (pathValue: string[], level: number) => Promise<T[]>;
  /**
   * @zh 下拉框收起展开时触发。
   * @en Callback when popup shown or hidden.
   */
  onVisibleChange?: (visible: boolean) => void;
  /**
   * @zh 点击清除时触发，参数是当前下拉框的展开状态。
   * @en Callback when click clear icon.
   */
  onClear?: (visible: boolean) => void;
}

export interface OptionProps {
  /**
   * @zh 选项的值
   * @en: the value of Option
   *
   */
  value?: string;
  /**
   * @zh 选项文本
   * @en option text
   */
  label?: string;
  /**
   * @zh 是否禁用该选项
   * @en whether to disabled
   */
  disabled?: boolean;
  /**
   * @zh 下一级选项
   * @en next level menu
   */
  children?: OptionProps[];
  /**
   * @zh 是否是叶子节点
   * @en Flag whether it is a leaf node
   */
  isLeaf?: boolean;
  /**
   * @zh 是否禁用复选框选中
   * @en Whether to disable the check box is selected
   * @version 2.21.0
   */
  disableCheckbox?: boolean;
  /**
   * @zh 其他字段
   * @en other fields
   */
  [key: string]: any;
}

/**
 * fieldnames 属性类型
 */
export type FieldNamesType = {
  /** 指定 label 在选项中对应的字段  */
  label?: string;
  /** 指定 value 在选项中对应的字段  */
  value?: string;
  /** 指定 children 在选项中对应的字段  */
  children?: string;
  /** 指定 disabled 在选项中对应的字段  */
  disabled?: string;
  /** 指定 isLeaf 在选项中对应的字段  */
  isLeaf?: string;
};

// 造成输入框值改变的原因：用户输入、选项下拉框收起、其他
export type InputValueChangeReason = 'manual' | 'optionListHide' | 'optionChecked';

export interface CascaderPanelProps<T> {
  className?: string | string[];
  style?: CSSProperties;
  store: Store<T>;
  multiple?: boolean;
  defaultValue?: string[][];
  value?: string[][];
  changeOnSelect?: boolean;
  popupVisible?: boolean;
  expandTrigger?: 'click' | 'hover';
  trigger?: 'click';
  prefixCls?: string;
  showEmptyChildren?: boolean;
  virtualListProps?: CascaderProps<T>['virtualListProps'];
  renderOption?: (option: NodeProps<T>, level: number) => ReactNode;
  onChange?: (value: string[][]) => void;
  loadMore?: (activeValue, level: number) => void;
  renderEmpty?: (width?: CSSProperties['width']) => ReactNode;
  renderFooter?: (level: number, activeOption: NodeProps<T> | null) => ReactNode;
  onDoubleClickOption?: () => void;
  onEsc?: () => void;
  dropdownColumnRender?: CascaderProps<T>['dropdownColumnRender'];
  dropdownMenuColumnStyle?: CascaderProps<T>['dropdownMenuColumnStyle'];
}
