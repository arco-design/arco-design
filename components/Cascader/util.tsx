import { isArray } from '../_util/is';
import Store from './base/store';
import { CascaderProps } from './interface';

export const ValueSeparator = '__arco_cascader__';

export const SHOW_PARENT = 'parent';
export const SHOW_CHILD = 'child';

export const PANEL_MODE = {
  cascader: 'cascader',
  select: 'select',
};

export function isEmptyValue(value) {
  return !value || (isArray(value) && value.length === 0);
}

export function getConfig(props: CascaderProps) {
  return {
    showEmptyChildren: props.showEmptyChildren,
    changeOnSelect: props.changeOnSelect,
    lazyload: !!props.loadMore,
    fieldNames: props.fieldNames,
    filterOption: props.filterOption,
    showParent:
      props.mode === 'multiple' && !props.changeOnSelect && props.checkedStrategy === SHOW_PARENT,
  };
}

export function getStore(props, value) {
  const tmp = value ? (Array.isArray(value[0]) ? value : [value]) : [];
  return new Store(props.options || [], tmp, getConfig(props));
}

export const transformValuesToSet = (values: string[][]) => {
  const _values = values || [];
  const valuesSet = _values.reduce((set, next) => {
    // 'next' could be a string.
    set.add([].concat(next).join(ValueSeparator));
    return set;
  }, new Set());

  return valuesSet;
};

export const valueInSet = (set, value: string[]) => {
  const _value = value || [];
  return set.has(_value.join(ValueSeparator));
};

export const removeValueFromSet = (set, value: string[]) => {
  const _value = value || [];
  return set.delete(_value.join(ValueSeparator));
};

export const formatValue = (value, isMultiple, store?): string[][] | undefined => {
  let _value = [];
  if (value === undefined) {
    _value = [];
  } else if (isMultiple) {
    _value = value;
  } else {
    _value = [value];
  }

  if (store && store.config.showParent) {
    const checkedNodes = store.getCheckedNodes();
    const valuesSet = transformValuesToSet(checkedNodes.map((node) => node.pathValue));
    const result = [];
    const temp = {};
    _value.map((v) => {
      v.some((_, index, arr) => {
        const curVal = arr.slice(0, index + 1);
        const pass = valueInSet(valuesSet, curVal);
        if (pass && !temp[curVal.join(ValueSeparator)]) {
          result.push(curVal);
          temp[curVal.join(ValueSeparator)] = 1;
        }
        return pass;
      });
    });

    return result;
  }
  return _value;
};

// change check status to false
const deny2Checked = (option) => {
  const deny = (options) => {
    return !Array.isArray(options)
      ? false
      : options.every((item) => {
          if (item._checked || item.disabled) {
            return true;
          }
          return deny(item.children);
        });
  };
  return option._halfChecked && deny(option?.children);
};

export const getMultipleCheckValue = (propsValue, store: Store<any>, option, _checked) => {
  const checked = _checked && deny2Checked(option) ? false : _checked;

  const beforeValueSet = store.getCheckedNodes().reduce((set, node) => {
    set.add(node.pathValue.join(ValueSeparator));
    return set;
  }, new Set());

  option.setCheckedState(checked);
  const checkedNodes = store.getCheckedNodes();
  const currentValue = checkedNodes.map((node) => node.pathValue);
  const currentValueSet = transformValuesToSet(currentValue);

  const newValueSet = new Set();
  return propsValue
    .filter((v) => {
      // v 不在 beforeValueSet 中，说明 v 不包含对应的option。直接返回true，不应该清除掉。
      if (!valueInSet(beforeValueSet, v) || valueInSet(currentValueSet, v)) {
        newValueSet.add(v.join(ValueSeparator));
        return true;
      }
    })
    .concat(
      currentValue.filter((v) => {
        return !valueInSet(newValueSet, v);
      })
    );
};
