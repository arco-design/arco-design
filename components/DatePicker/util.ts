import dayjs, { Dayjs } from 'dayjs';
import { isArray, isDayjs } from '../_util/is';
import { methods } from '../_util/dayjs';

function getFormat(time) {
  return isDayjs(time) && time.format('HH:mm:ss');
}

export function isTimeArrayChange(prevTime: Dayjs[], nextTime: Dayjs[]) {
  return (
    getFormat(prevTime[0]) !== getFormat(nextTime[0]) ||
    getFormat(prevTime[1]) !== getFormat(nextTime[1])
  );
}

export function getAvailableDayjsLength(value) {
  if (!value) {
    return 0;
  }
  if (isArray(value)) {
    if (isDayjs(value[0]) && isDayjs(value[1])) {
      return 2;
    }
    if (!isDayjs(value[0]) && !isDayjs(value[1])) {
      return 0;
    }
    return 1;
  }
  return 0;
}

// https://github.com/react-component/picker/blob/master/src/utils/dateUtil.ts#L234
export function isDisabledDate(cellDate, disabledDate, mode): boolean {
  if (typeof disabledDate !== 'function') {
    return false;
  }
  // Whether cellDate is disabled in range
  const getDisabledFromRange = (
    currentMode: 'date' | 'month' | 'year',
    start: number,
    end: number
  ) => {
    let current = start;
    while (current <= end) {
      let date: Dayjs;
      switch (currentMode) {
        case 'date': {
          date = methods.set(cellDate, 'date', current);
          if (!disabledDate(date)) {
            return false;
          }
          break;
        }
        case 'month': {
          date = methods.set(cellDate, 'month', current);
          if (!isDisabledDate(date, disabledDate, 'month')) {
            return false;
          }
          break;
        }
        case 'year': {
          date = methods.set(cellDate, 'year', current);
          if (!isDisabledDate(date, disabledDate, 'year')) {
            return false;
          }
          break;
        }
        default:
          break;
      }
      current += 1;
    }
    return true;
  };
  switch (mode) {
    case 'date':
    case 'week': {
      return disabledDate(cellDate);
    }
    case 'month': {
      const startDate = 1;
      const endDate = cellDate.endOf('month').get('date');
      return getDisabledFromRange('date', startDate, endDate);
    }
    case 'quarter': {
      const startMonth = Math.floor(cellDate.get('month') / 3) * 3;
      const endMonth = startMonth + 2;
      return getDisabledFromRange('month', startMonth, endMonth);
    }
    case 'year': {
      return getDisabledFromRange('month', 0, 11);
    }
    default:
      return false;
  }
}

type WeekStartType = 0 | 1 | 2 | 3 | 4 | 5 | 6;

export function getDefaultWeekStart(dayjsLocale: string): WeekStartType {
  return (dayjs.Ls?.[dayjsLocale]?.weekStart as WeekStartType) || 0;
}

export function getLocaleDayjsValue(
  date: Dayjs | undefined,
  dayjsLocale: string
): Dayjs | undefined {
  return date ? date.locale(dayjsLocale) : date;
}
