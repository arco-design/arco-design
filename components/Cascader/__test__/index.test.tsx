import React from 'react';
import { act } from 'react-test-renderer';
import mountTest from '../../../tests/mountTest';
import componentConfigTest from '../../../tests/componentConfigTest';
import Cascader from '../cascader';
import { fireEvent, render } from '../../../tests/util';

mountTest(Cascader);
componentConfigTest(Cascader, 'Cascader');

const prefixCls = '.arco-cascader';
const options = [
  {
    value: 'shanghai',
    label: '上海',
    children: [
      {
        value: 'shanghaishi',
        label: '上海市',
      },
    ],
  },
];

function mountCascader(component: React.ReactElement) {
  return render(component);
}

// 模块

describe('Cascader basic test', () => {
  // 生命周期
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    // 运行所有定时器
    jest.runAllTimers();
  });

  it('popupChange correctly', () => {
    const mockFn = jest.fn();
    const wrapper = mountCascader(<Cascader options={options} onVisibleChange={mockFn} />);
    // 找到class为.arco-cascader dom
    fireEvent.click(wrapper.find(prefixCls)[0]);
    expect(mockFn).toHaveBeenCalledWith(true); // 被调用，传入参数为true
    // 展开第一级
    expect(wrapper.find(`${prefixCls}-list-column`)).toHaveLength(1);
    fireEvent.click(wrapper.find(prefixCls)[0]);
    // 消失有动画的延时
    jest.runAllTimers();
    expect(mockFn).toHaveBeenCalledWith(false); // 传入参数false
  });

  it('onKeyDown is called', () => {
    const onKeyDown = jest.fn();
    const wrapper = mountCascader(<Cascader onKeyDown={onKeyDown} />);
    fireEvent.keyDown(wrapper.querySelector('input'));
    expect(onKeyDown).toHaveBeenCalled();
  });

  it('select correctly', async () => {
    const mockChange = jest.fn();
    const mockFn = jest.fn();

    const wrapper = mountCascader(
      <Cascader onVisibleChange={mockFn} options={options} onChange={mockChange} />
    );
    // 展开一层
    fireEvent.click(wrapper.find(prefixCls)[0]);
    const list1 = wrapper.querySelector(`${prefixCls}-list-column`);
    // 展开第二层
    fireEvent.click(
      list1.querySelector(`${prefixCls}-list-item ${prefixCls}-list-item-label`) as Element
    );
    expect(wrapper.find(`${prefixCls}-list-column`)).toHaveLength(2);

    const list2 = wrapper.find(`${prefixCls}-list-column`);

    fireEvent.click(
      list2[list2.length - 1].querySelector(`${prefixCls}-list-item ${prefixCls}-list-item-label`)
    );
    // onChange事件触发
    expect(mockChange).toBeCalledTimes(1); // 得到mock函数被触发的次数
    expect(mockChange.mock.calls[0][0]).toEqual(['shanghai', 'shanghaishi']);
    expect(mockChange.mock.calls[0][1].map((item) => item.value)).toEqual([
      'shanghai',
      'shanghaishi',
    ]);

    expect(wrapper.find('input')[0].getAttribute('value')).toBe('上海 / 上海市');
    // 消失有动画的延时
    jest.runAllTimers();
    expect(mockFn).toHaveBeenCalledWith(false); // 传入参数false
  });

  it('changeOnselect correctly', () => {
    const wrapper = mountCascader(<Cascader options={options} changeOnSelect />);
    // 展开一层
    fireEvent.click(wrapper.find(prefixCls)[0]);
    const list1 = wrapper.querySelector(`${prefixCls}-list-column`);
    fireEvent.click(
      list1.querySelector(`${prefixCls}-list-item ${prefixCls}-list-item-label`) as Element
    );
    expect(wrapper.querySelector('input').getAttribute('value')).toBe('上海');
  });
});

let wrapper;
describe('support clear correctly', () => {
  beforeEach(() => {
    const defaultValue = ['shanghai', 'shanghaishi'];
    wrapper = mountCascader(<Cascader options={options} defaultValue={defaultValue} allowClear />);
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runAllTimers();
  });

  // 默认选中值正确
  it('set defaultValue correctly', () => {
    expect(wrapper.querySelector('input').getAttribute('value')).toBe('上海 / 上海市');
    fireEvent.click(wrapper.find(prefixCls)[0]);

    expect(wrapper.find('.arco-cascader-list-column')).toHaveLength(2);
  });

  // 清除值正确
  it('clear value correctly', () => {
    fireEvent.click(wrapper.find('.arco-icon-close')[0]);
    expect(wrapper.find(`${prefixCls}-view`)[0].textContent).toBe('');
    fireEvent.click(wrapper.find(prefixCls)[0]);
  });
});

describe('support multiple correctly', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runAllTimers();
  });
  const options = [
    {
      value: 'beijing',
      label: '北京',
      children: [
        {
          value: 'beijingshi',
          label: '北京市',
          children: [
            {
              value: 'chaoyang',
              label: '朝阳区',
              children: [
                {
                  value: 'datunli',
                  label: '大屯里',
                },
              ],
            },
            {
              value: 'dongcheng',
              label: '东城区',
            },
            {
              value: 'xicheng',
              label: '西城区',
            },
            {
              value: 'haidian',
              label: '海淀区',
            },
          ],
        },
      ],
    },
  ];

  it('multiply select correctly', () => {
    let value = [['beijing', 'beijingshi', 'chaoyang', 'datunli']];
    let option;
    const wrapper = mountCascader(
      <Cascader
        key="1"
        placeholder="请选择一个地点"
        style={{ maxWidth: 300 }}
        options={options}
        onChange={(v, o) => {
          value = v as string[][];
          option = o;
        }}
        mode="multiple"
        defaultValue={value}
      />
    );
    fireEvent.click(wrapper.find(prefixCls)[0]);
    jest.runAllTimers();

    expect(wrapper.find(`${prefixCls}-list-column`)).toHaveLength(4);
    const list = wrapper.find(`${prefixCls}-list-column`)[2];

    act(() => {
      fireEvent.click(list.querySelector(`${prefixCls}-list-item .arco-checkbox `), {});
    });
    jest.runAllTimers();
    expect(value).toEqual([]);
    act(() => {
      fireEvent.click(
        list.querySelectorAll(`${prefixCls}-list-item`).item(1).querySelector('.arco-checkbox')
      );
      fireEvent.click(
        list.querySelectorAll(`${prefixCls}-list-item`).item(2).querySelector('.arco-checkbox')
      );
    });

    expect(value.length).toEqual(2);
    expect(option.length).toEqual(2);
  });

  it('multiply select checkedStrategy correctly controlled', () => {
    let value = [['beijing']];
    const wrapper = mountCascader(
      <Cascader
        renderFormat={(values) => values.join('-')}
        placeholder="请选择一个地点"
        style={{ maxWidth: 300 }}
        options={options}
        onChange={(v) => {
          value = v as string[][];
        }}
        checkedStrategy="parent"
        mode="multiple"
        value={value}
      />
    );
    expect(wrapper.find('.arco-tag > span')[0].textContent).toBe(`北京`);

    fireEvent.click(wrapper.find(prefixCls)[0]);
    jest.runAllTimers();

    fireEvent.click(
      wrapper.find(`${prefixCls}-list-column`)[0].querySelector(`${prefixCls}-list-item-label`)
    );
    jest.runAllTimers();
    fireEvent.click(
      wrapper.find(`${prefixCls}-list-column`)[1].querySelector(`${prefixCls}-list-item-label`)
    );

    jest.runAllTimers();

    expect(wrapper.find('.arco-checkbox').length).toEqual(
      wrapper.find('.arco-checkbox-checked').length
    );

    expect(wrapper.find(`${prefixCls}-list-column`)).toHaveLength(3);

    jest.runAllTimers();

    fireEvent.click(
      wrapper
        .find(`${prefixCls}-list-column`)[2]
        .querySelectorAll(`${prefixCls}-list-item`)[1]
        .querySelector('.arco-checkbox')
    );

    jest.runAllTimers();
    expect(value).toEqual([
      ['beijing', 'beijingshi', 'chaoyang'],
      ['beijing', 'beijingshi', 'xicheng'],
      ['beijing', 'beijingshi', 'haidian'],
    ]);
  });

  it('multiply select checkedStrategy correctly', () => {
    let value = [['beijing', 'beijingshi', 'chaoyang', 'datunli']];
    const wrapper = mountCascader(
      <Cascader
        renderFormat={(values) => values.join('-')}
        placeholder="请选择一个地点"
        style={{ maxWidth: 300 }}
        options={options}
        onChange={(v) => {
          value = v as string[][];
        }}
        checkedStrategy="parent"
        mode="multiple"
        defaultValue={value}
      />
    );
    expect(wrapper.find('.arco-tag > span').item(0).textContent).toBe(`北京-北京市-朝阳区`);

    fireEvent.click(wrapper.container.querySelector(prefixCls) as Element);
    jest.runAllTimers();
    fireEvent.click(document.querySelector('.arco-checkbox') as Element);

    expect(value).toEqual([['beijing']]);
  });

  it('support multiply search correctly', () => {
    // const wrapper = mountCascader(<Cascader  options={options} showSearch mode="multiple" />);
    // fireEvent.click(wrapper.find(prefixCls)[0]);
    // wrapper.find(`${prefixCls}-view-search>input`).simulate('change', {
    //   target: {
    //     value: '北京',
    //   },
    // });
    // const store = wrapper.find('Cascader').instance().store;
    // const mockSearch = jest.spyOn(store, 'searchNodeByLabel');
  });

  describe('loadmore', () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.runAllTimers();
    });

    const options = [
      {
        value: 'beijing',
        label: '北京',
      },
      {
        value: 'shanghai',
        label: '上海',
      },
    ];
    const mockfn = jest.fn();
    it('loadmore correctly', () => {
      const wrapper = mountCascader(
        <Cascader
          style={{ maxWidth: 300 }}
          options={options}
          loadMore={mockfn}
          showSearch
          mode="multiple"
        />
      );
      fireEvent.click(wrapper.find(prefixCls)[0]);
      const list = wrapper.find(`${prefixCls}-list-column`)[0];
      fireEvent.click(
        list.querySelectorAll(`${prefixCls}-list-item ${prefixCls}-list-item-label`).item(0)
      );
      expect(mockfn.mock.calls).toHaveLength(1);
    });

    it('value control', () => {
      // const Demo = function() {
      //   const [value, setValue] = React.useState([]);
      //   return (
      //     <Cascader
      //       placeholder="请选择一个地点"
      //       options={options}
      //       mode="multiple"
      //       value={value}
      //       onChange={setValue}
      //     />
      //   );
      // };
      // const wrapper = mountCascader(<Demo />);
      // const cascader = wrapper.find('Cascader');
      // cascader.find(prefixCls).simulate('click');
      // const list1 = cascader.find(`${prefixCls}-list-column`);
      // list1.find(`${prefixCls}-list-item ${prefixCls}-list-item-label`).simulate('click');
    });
  });
});

describe('fieldNames Cascader', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.runAllTimers();
  });
  const options = [
    {
      id: 'beijing',
      name: '北京',
      child: [
        {
          id: 'beijingshi',
          name: '北京市',
          child: [
            {
              id: 'chaoyang',
              name: '朝阳区',
              child: [
                {
                  id: 'datunli',
                  name: '大屯里',
                },
              ],
            },
            {
              id: 'dongcheng',
              name: '东城区',
            },
            {
              id: 'xicheng',
              name: '西城区',
            },
            {
              id: 'haidian',
              name: '海淀区',
            },
          ],
        },
      ],
    },
  ];

  it('multiply select correctly', () => {
    let value;
    const wrapper = mountCascader(
      <Cascader
        key="multiplyselect"
        fieldNames={{
          label: 'name',
          value: 'id',
          children: 'child',
        }}
        placeholder="请选择一个地点"
        onChange={(v) => {
          value = v;
        }}
        style={{ maxWidth: 300 }}
        options={options}
        mode="multiple"
        defaultValue={[['beijing', 'beijingshi', 'chaoyang', 'datunli']]}
      />
    );
    expect(wrapper.find('.arco-tag').length).toBe(1);
    act(() => {
      fireEvent.click(wrapper.find(prefixCls)[0]);
    });

    expect(wrapper.find(`.arco-cascader-list-column`)).toHaveLength(4);
    const list = wrapper.find(`${prefixCls}-list-column`).item(2);

    act(() => {
      fireEvent.click(
        list.querySelectorAll(`${prefixCls}-list-item`).item(0).querySelector('.arco-checkbox')
      );
    });

    expect(value.length).toBe(0);
    act(() => {
      fireEvent.click(
        list.querySelectorAll(`${prefixCls}-list-item`).item(1).querySelector('.arco-checkbox')
      );
    });

    act(() => {
      fireEvent.click(
        list.querySelectorAll(`${prefixCls}-list-item`).item(2).querySelector('.arco-checkbox')
      );
    });
    expect(value.length).toBe(2);
  });

  describe('loadmore', () => {
    beforeEach(() => {
      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.runAllTimers();
    });

    const options = [
      {
        id: 'beijing',
        name: '北京',
      },
      {
        id: 'shanghai',
        name: '上海',
      },
    ];
    const mockfn = jest.fn();
    it('loadmore correctly', () => {
      const wrapper = mountCascader(
        <Cascader
          fieldNames={{ label: 'name', isLeaf: 'is-leaf', value: 'id' }}
          style={{ maxWidth: 300 }}
          options={options}
          loadMore={mockfn}
          showSearch
          mode="multiple"
        />
      );
      fireEvent.click(wrapper.find(prefixCls)[0]);
      const list = wrapper.find(`${prefixCls}-list-column`)[0];
      fireEvent.click(
        list.querySelectorAll(`${prefixCls}-list-item ${prefixCls}-list-item-label`).item(0)
      );
      expect(mockfn.mock.calls).toHaveLength(1);
    });
  });
});
