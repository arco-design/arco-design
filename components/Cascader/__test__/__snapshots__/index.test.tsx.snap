// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider componentConfig.Cascader default 1`] = `
<div
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="listbox"
  class="arco-cascader arco-cascader-single arco-cascader-size-default"
  role="combobox"
  tabindex="0"
>
  <div
    class="arco-cascader-view"
    title=""
  >
    <span
      class="arco-cascader-view-selector"
    >
      <input
        autocomplete="off"
        class="arco-cascader-view-input"
        style="width:100%;pointer-events:none"
        tabindex="-1"
        value=""
      />
      <span
        class="arco-cascader-view-value arco-cascader-view-value-mirror"
      />
    </span>
    <div
      aria-hidden="true"
      class="arco-cascader-suffix"
    >
      <div
        class="arco-cascader-arrow-icon"
      >
        <svg
          aria-hidden="true"
          class="arco-icon arco-icon-down"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 48 48"
        >
          <path
            d="M39.6 17.443 24.043 33 8.487 17.443"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Cascader set className globally 1`] = `
<div
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="listbox"
  class="arco-cascader arco-cascader-single arco-cascader-size-default global-default-class"
  role="combobox"
  tabindex="0"
>
  <div
    class="arco-cascader-view"
    title=""
  >
    <span
      class="arco-cascader-view-selector"
    >
      <input
        autocomplete="off"
        class="arco-cascader-view-input"
        style="width:100%;pointer-events:none"
        tabindex="-1"
        value=""
      />
      <span
        class="arco-cascader-view-value arco-cascader-view-value-mirror"
      />
    </span>
    <div
      aria-hidden="true"
      class="arco-cascader-suffix"
    >
      <div
        class="arco-cascader-arrow-icon"
      >
        <svg
          aria-hidden="true"
          class="arco-icon arco-icon-down"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 48 48"
        >
          <path
            d="M39.6 17.443 24.043 33 8.487 17.443"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Cascader set className globally and component set className itself. 1`] = `
<div
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="listbox"
  class="arco-cascader arco-cascader-single arco-cascader-size-default component-class"
  role="combobox"
  tabindex="0"
>
  <div
    class="arco-cascader-view"
    title=""
  >
    <span
      class="arco-cascader-view-selector"
    >
      <input
        autocomplete="off"
        class="arco-cascader-view-input"
        style="width:100%;pointer-events:none"
        tabindex="-1"
        value=""
      />
      <span
        class="arco-cascader-view-value arco-cascader-view-value-mirror"
      />
    </span>
    <div
      aria-hidden="true"
      class="arco-cascader-suffix"
    >
      <div
        class="arco-cascader-arrow-icon"
      >
        <svg
          aria-hidden="true"
          class="arco-icon arco-icon-down"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 48 48"
        >
          <path
            d="M39.6 17.443 24.043 33 8.487 17.443"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`ConfigProvider componentConfig.Cascader set data-* property 1`] = `
<div
  aria-autocomplete="list"
  aria-expanded="false"
  aria-haspopup="listbox"
  class="arco-cascader arco-cascader-single arco-cascader-size-default"
  data-test-name="Cascader"
  role="combobox"
  tabindex="0"
>
  <div
    class="arco-cascader-view"
    title=""
  >
    <span
      class="arco-cascader-view-selector"
    >
      <input
        autocomplete="off"
        class="arco-cascader-view-input"
        style="width:100%;pointer-events:none"
        tabindex="-1"
        value=""
      />
      <span
        class="arco-cascader-view-value arco-cascader-view-value-mirror"
      />
    </span>
    <div
      aria-hidden="true"
      class="arco-cascader-suffix"
    >
      <div
        class="arco-cascader-arrow-icon"
      >
        <svg
          aria-hidden="true"
          class="arco-icon arco-icon-down"
          fill="none"
          focusable="false"
          stroke="currentColor"
          stroke-width="4"
          viewBox="0 0 48 48"
        >
          <path
            d="M39.6 17.443 24.043 33 8.487 17.443"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
